### PR TITLE
[PHP 5.5] Remove extra preslash in class reference, as in practice removal manually

### PR DESCRIPTION
**Pull request by TomasVotruba**
I've learned in Mautic upgrade sprint, this rule is adding extra slash that has to be removed.
Let's make it easy to use :+1:

### CODE REVIEW

File: rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc [hunk 1]
```diff
@@ -16,11 +16,11 @@ final class PreSlash
 
 namespace Rector\Tests\Php55\Rector\String_\StringClassNameToClassConstantRector\Fixture;
 
-final class PreSlash
+final class NoPreSlash
 {
     public function preSlash()
     {
-        return '\\' . \Rector\Tests\Php55\Rector\String_\StringClassNameToClassConstantRector\Source\AnotherClass::class;
+        return \Rector\Tests\Php55\Rector\String_\StringClassNameToClassConstantRector\Source\AnotherClass::class;
```
Comment (samsonasik) [inline rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: This is invalid, the result is different, the comparison printed  first has pre-backslash, the later it don't have as using ::class

Comment (samsonasik) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: See why it is invalid

https://3v4l.org/I947K

Comment (TomasVotruba) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: Yes, it's different, but in practise this value is prefered. The class must exist and does in both cases.

Comment (samsonasik) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: I can remember that in CI4, that string controller namespace use that comparison, which now will invalid, /cc @kenjis

Comment (TomasVotruba) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: The CI4 code should be update then to use the class name, not require a slash.  Seems like a hack :)
 
In the Mautic case it added slash that made code less readable.

Comment (kenjis) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: It is a special rule for CI4 routing. It distinguishes between `Foo` (relative to the default namespace) and `\Foo` (FQCN).
I know it is not common usage, but we cannot change. If we change, the existing apps will break.

Comment (TomasVotruba) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: No need to change routing.
I think that's rather edge case compared to common use of `::class`, so we'll stick with behavior above as default.

Comment (samsonasik) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: @TomasVotruba how about make it configurable, eg, add config flag `keepBackslash` ?

Comment (leoloso) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: This:

> It distinguishes between Foo (relative to the default namespace) and \Foo (FQCN).

They are not the same. For instance, you could have classes `SomeNamespace\StringUtils` and `StringUtils`. Then, referencing `StringUtils` or `\StringUtils` from `SomeNamespace\ObjectUtils` will be referencing 2 different classes.

So I think that you'd rather keep that pre-backlash if you're targeting the class on the root, or it may produce bugs.

Comment (staabm) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: I agree with leoloso.

maybe we can detect the mautic case somehow instead of removing the slash in the general case?

Comment (TomasVotruba) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: @staabm I had similar issues with othe projects that we had to remove manually.

Could you share link to Github piece of code where string is used and extra slash is relevant?

Comment (TomasVotruba) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: @leoloso 

> They are not the same. For instance, you could have classes SomeNamespace\StringUtils and StringUtils. Then, referencing StringUtils or \StringUtils from SomeNamespace\ObjectUtils will be referencing 2 different classes.

This is different use case. I talk about concatted slash string:

```diff
-'\\' . SomeClass::class
+SomeClass::class
```

Two same named classes are kept separated.

Comment (kenjis) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: Probably the following test will break with this PR.
https://github.com/codeigniter4/CodeIgniter4/pull/8321/commits/7646e3fc97a9c99fb12133eb29f8a72832642d02

Comment (samsonasik) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: @kenjis yes, that will break that

Comment (samsonasik) [reply rules-tests/Php55/Rector/String_/StringClassNameToClassConstantRector/Fixture/no_pre_slash.php.inc:23]: I create new PR to make it configurable:

- https://github.com/rectorphp/rector-src/pull/5354
